pkg/pgxconn/helpers.go

```diff
@@ -0,0 +1,178 @@
+
+package pgxconn
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/jackc/pgx/v4"
+)
+
+const (
+	timeoutDisabled = 0
+	minTimeoutMs    = 1
+)
+
+type rowsFromSingleQueryBatch struct {
+	PgxRows
+	batchResults pgx.BatchResults
+}
+
+func (b *rowsFromSingleQueryBatch) Close() {
+	b.PgxRows.Close()
+	_ = b.batchResults.Close()
+}
+
+// StatementTimeoutFromCtx returns the time in milliseconds remaining
+// before the context expires to be used in a `statement_timeout`.
+//
+// If the context had a deadline set but the remaining time is lower than
+// zero, a 1 Ms timeout value will be returned instead. This is because
+// zero disables statement_timeout and it's also the minimum allowed value
+// accepted by the function. Either way, at this point the context is
+// most likely canceled and the request won't made.
+func StatementTimeoutFromCtx(ctx context.Context) (int64, bool) {
+	t, ok := ctx.Deadline()
+	if !ok {
+		return timeoutDisabled, false
+	}
+	d := time.Until(t).Milliseconds()
+	if d < minTimeoutMs {
+		return minTimeoutMs, true
+	}
+	return d, true
+}
+
+// QueryWithTimeout wraps the given query in a transaction that has
+// statement_timeout set to the given timeout value. The timeout value is
+// taken as milliseconds. A timeout less than 0 will disable the timeout.
+//
+// The returned `commit` functions returns the error, if any,
+// from the Exec("COMMIT") part of the transaction. Since this is the
+// final query from the underlying batch, subsequent calls will yield then
+// `batch already close` error.
+//
+// For more contect. Taken from the postgreSQL docs:
+// https://www.postgresql.org/docs/current/runtime-config-client.html
+//
+// statement_timeout (integer)
+//
+// Abort any statement that takes more than the specified amount of time. If
+// log_min_error_statement is set to ERROR or lower, the statement that timed
+// out will also be logged. If this value is specified without units, it is
+// taken as milliseconds. A value of zero (the default) disables the timeout.
+//
+// The timeout is measured from the time a command arrives at the server until
+// it is completed by the server. If multiple SQL statements appear in a single
+// simple-Query message, the timeout is applied to each statement separately.
+// (PostgreSQL versions before 13 usually treated the timeout as applying to
+// the whole query string.) In extended query protocol, the timeout starts
+// running when any query-related message (Parse, Bind, Execute, Describe)
+// arrives, and it is canceled by completion of an Execute or Sync message.
+//
+// Setting statement_timeout in postgresql.conf is not recommended because it
+// would affect all sessions.
+func QueryWithTimeout(
+	ctx context.Context,
+	conn PgxConn,
+	timeout int64,
+	sql string,
+	args ...interface{},
+) (rows PgxRows, commit func() error, err error) {
+	noopCommit := func() error { return nil }
+	if timeout < timeoutDisabled {
+		timeout = timeoutDisabled
+	}
```

Comment:
In postgres, a statement_timeout = 0 disables the timeout, but here we use < 0. I would prefer that we match postgres so that it's less confusing to people familiar with how statement_timeout works. Is there a reason we used < 0 here instead of = 0?